csi/controller.go

```diff
@@ -0,0 +1,230 @@
+/*
+Package csi is CSI driver interface for OSD
+Copyright 2017 Portworx
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package csi
+
+import (
+	"fmt"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"go.pedge.io/dlog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ControllerGetCapabilities is a CSI API functions which returns to the caller
+// the capabilities of the OSD CSI driver.
+func (s *OsdCsiServer) ControllerGetCapabilities(
+	ctx context.Context,
+	req *csi.ControllerGetCapabilitiesRequest) (*csi.ControllerGetCapabilitiesResponse, error) {
+
+	// Volume capabilities
+	capCreateDeleteVolume := &csi.ControllerServiceCapability{
+		Type: &csi.ControllerServiceCapability_Rpc{
+			Rpc: &csi.ControllerServiceCapability_RPC{
+				Type: csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME,
+			},
+		},
+	}
+
+	//
+	// Here we will be adding the following in future patches:
+	// List Volumes Capabability
+	// Get Storage Capacity
+	//
+
+	return &csi.ControllerGetCapabilitiesResponse{
+		Reply: &csi.ControllerGetCapabilitiesResponse_Result_{
+			Result: &csi.ControllerGetCapabilitiesResponse_Result{
+				Capabilities: []*csi.ControllerServiceCapability{
+					capCreateDeleteVolume,
+				},
+			},
+		},
+	}, nil
+
+}
+
+// ControllerPublishVolume is a CSI API implements the attachment of a volume
+// on to a node.
+func (s *OsdCsiServer) ControllerPublishVolume(
+	context.Context,
+	*csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "This request is not supported")
```

Comment:
Is the CSI client expecting this error code as the error for No-op?

Reply:
This is for clients that are incorrectly written

Reply:
discussed offline with @lpabon